src/compiler/driver.jl

```diff
@@ -1,90 +1,200 @@
+        globalUnique = previous_globalUnique
+    end
+
+    return codegen(target, job;
+                   libraries=libraries,
+                   optimize=optimize, strip=strip, strict=strict)
 end
 
-function compile(ctx::CompilerContext)
-    if compile_hook[] != nothing
-        hook = compile_hook[]
-        compile_hook[] = nothing
+function codegen(target::Symbol, job::CompilerJob;
+                 libraries::Bool=true, optimize::Bool=true,
+                 strip::Bool=false, strict::Bool=true)
+    ## Julia IR
+
+    @timeit to[] "validation" check_method(job)
+
+    @timeit to[] "Julia front-end" begin
+
+        # get the method instance
+        world = typemax(UInt)
+        meth = which(job.f, job.tt)
+        sig = Base.signature_type(job.f, job.tt)::Type
+        (ti, env) = ccall(:jl_type_intersection_with_env, Any,
+                          (Any, Any), sig, meth.sig)::Core.SimpleVector
+        if VERSION >= v"1.2.0-DEV.320"
+            meth = Base.func_for_method_checked(meth, ti, env)
+        else
+            meth = Base.func_for_method_checked(meth, ti)
+        end
+        method_instance = ccall(:jl_specializations_get_linfo, Ref{Core.MethodInstance},
+                      (Any, Any, Any, UInt), meth, ti, env, world)
 
-        global globalUnique
-        previous_globalUnique = globalUnique
+        for var in env
+            if var isa TypeVar
+                throw(KernelError(job, "method captures a typevar (you probably use an unbound type variable)"))
+            end
+        end
+    end
 
-        hook(ctx)
+    target == :julia && return method_instance
 
-        globalUnique = previous_globalUnique
-        compile_hook[] = hook
+
+    ## LLVM IR
+
+    defs(mod)  = filter(f -> !isdeclaration(f), collect(functions(mod)))
+    decls(mod) = filter(f ->  isdeclaration(f) && intrinsic_id(f) == 0,
+                        collect(functions(mod)))
+    need_library(mod, lib) = any(f -> isdeclaration(f) &&
+                                      intrinsic_id(f) == 0 &&
+                                      haskey(functions(lib), LLVM.name(f)),
+                                 functions(mod))
+
+    # always preload the runtime, and do so early; it cannot be part of any timing block
+    # because it recurses into the compiler
+    if libraries
+        runtime = load_runtime(job.agent)
+        runtime_fns = LLVM.name.(defs(runtime))
     end
 
+    @timeit to[] "LLVM middle-end" begin
+        ir, kernel = @timeit to[] "IR generation" irgen(job, method_instance, world)
+
+        if libraries
+            undefined_fns = LLVM.name.(decls(ir))
+            #= FIXME: Remove this "__nv_" check
+            #
+            if any(fn->startswith(fn, "__nv_"), undefined_fns)
+                libdevice = load_libdevice(job.cap)
+                @timeit to[] "device library" link_libdevice!(job, ir, libdevice)
+            end
+            =#
+            # FIXME: Load this only when needed
```

Comment:
As long as linking them doesn't cause the performance issues (like on nvidia) doing this unconditionally should be fine.